In [80]:
import numpy as np
import pandas as pd

In [81]:
data = pd.read_csv('data/2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv', index_col='order_id')

In [82]:
data.tail(25)

shop_id  user_id  order_amount  total_items payment_method  \
order_id                                                               
4976           76      932           155            1           cash   
4977           93      763           114            1    credit_card   
4978           69      903           131            1    credit_card   
4979            2      841           282            3           cash   
4980           75      820           128            1          debit   
4981           50      942           772            4    credit_card   
4982           86      715           260            2          debit   
4983            1      906           316            2    credit_card   
4984           46      969           332            2           cash   
4985           44      966           432            3    credit_card   
4986            8      916           396            3          debit   
4987          100      731           111            1           cash   
4988           11      826           184            1    credit_card   
4989           86      877           260            2           cash   
4990           11      843           552            3    credit_card   
4991           24      860           140            1           cash   
4992           61      707           158            1           cash   
4993           49      739           258            2          debit   
4994            4      834           384            3          debit   
4995           12      954           201            1           cash   
4996           73      993           330            2          debit   
4997           48      789           234            2           cash   
4998           56      867           351            3           cash   
4999           60      825           354            2    credit_card   
5000           44      734           288            2          debit   

                   created_at  
order_id                       
4976       2017-03-25 1:15:09  
4977      2017-03-20 17:14:51  
4978      2017-03-13 10:39:49  
4979      2017-03-09 19:13:23  
4980      2017-03-25 12:28:18  
4981       2017-03-23 5:15:19  
4982      2017-03-10 18:53:38  
4983      2017-03-26 15:57:18  
4984      2017-03-23 19:18:15  
4985      2017-03-15 20:12:31  
4986      2017-03-15 14:22:51  
4987       2017-03-30 7:01:59  
4988       2017-03-02 9:35:25  
4989      2017-03-12 19:55:36  
4990      2017-03-02 15:06:29  
4991       2017-03-26 0:32:39  
4992      2017-03-13 18:56:04  
4993      2017-03-24 13:48:08  
4994       2017-03-18 4:12:20  
4995       2017-03-22 0:38:11  
4996      2017-03-30 13:47:17  
4997      2017-03-16 20:36:16  
4998       2017-03-19 5:42:42  
4999      2017-03-16 14:51:18  
5000      2017-03-18 15:48:18

In [83]:
data['order_amount'].sum()/ 5000

3145.128

In [91]:
data['order_amount'].max()

704000

The average number of orders in the store is 50


In [84]:
data.head()

shop_id  user_id  order_amount  total_items payment_method  \
order_id                                                               
1              53      746           224            2           cash   
2              92      925            90            1           cash   
3              44      861           144            1           cash   
4              18      935           156            1    credit_card   
5              18      883           156            1    credit_card   

                   created_at  
order_id                       
1         2017-03-13 12:36:56  
2         2017-03-03 17:38:52  
3          2017-03-14 4:23:56  
4         2017-03-26 12:43:37  
5          2017-03-01 4:35:11

In [85]:
data.mean(axis= 0)

/var/folders/xg/x9dp1h1x32q36fc2kpxyxd180000gn/T/ipykernel_28750/4034443637.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data.mean(axis= 0)


shop_id           50.0788
user_id          849.0924
order_amount    3145.1280
total_items        8.7872
dtype: float64

In [86]:
data.std(axis=0)

/var/folders/xg/x9dp1h1x32q36fc2kpxyxd180000gn/T/ipykernel_28750/1215420905.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data.std(axis=0)


shop_id            29.006118
user_id            87.798982
order_amount    41282.539349
total_items       116.320320
dtype: float64

### Average per unit order value

In [87]:
data["unit_order_amount"] = data["order_amount"]/data["total_items"]

In [88]:
data.head()

shop_id  user_id  order_amount  total_items payment_method  \
order_id                                                               
1              53      746           224            2           cash   
2              92      925            90            1           cash   
3              44      861           144            1           cash   
4              18      935           156            1    credit_card   
5              18      883           156            1    credit_card   

                   created_at  unit_order_amount  
order_id                                          
1         2017-03-13 12:36:56              112.0  
2         2017-03-03 17:38:52               90.0  
3          2017-03-14 4:23:56              144.0  
4         2017-03-26 12:43:37              156.0  
5          2017-03-01 4:35:11              156.0

In [89]:
data.mean()

/var/folders/xg/x9dp1h1x32q36fc2kpxyxd180000gn/T/ipykernel_28750/531903386.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data.mean()


shop_id                50.0788
user_id               849.0924
order_amount         3145.1280
total_items             8.7872
unit_order_amount     387.7428
dtype: float64

In [90]:
data.std()

/var/folders/xg/x9dp1h1x32q36fc2kpxyxd180000gn/T/ipykernel_28750/2723740006.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data.std()


shop_id                 29.006118
user_id                 87.798982
order_amount         41282.539349
total_items            116.320320
unit_order_amount     2441.963725
dtype: float64

### Q1

#### 1. Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 

The AOV of 3145.13 dollars was most likely gotten by just taking the mean of the order_amount column. This is problematic because of outliers like 704000 dollars. This user could be buying for wholesale purposes and would not reflect the buying behaviour of regular sneaker purchases. The Standard Deviation of the Order Amount is 41282.54 dollars, this shows that most of the values are very far from the average. Thus while the average was calculated properly, it does not give an accurate depiction of the average buyer. 

#### 2. What metric would you report for this dataset?

A better metric would be the Unit Order Amount. The largest Order Amount came from an Order of 2000 items, this evaluates to a unit order amount of 352 dollars. The Unit Order Amount scales the amount from each buyer, thus providing a more accurate depiction of the average buyer. It also has a much smaller standard deviation of 2441.96 dollars.

#### 3. What is its value?

The Unit Order Amount is 387.74 dollars


### Q2
#### 1. How many orders were shipped by Speedy Express in total?

SELECT * FROM [Orders]
WHERE ShipperID == 1

Ans = 54


#### 2. What is the last name of the employee with the most orders?

SELECT LASTNAME, MAX(NOrders)
FROM (SELECT Orders.EmployeeID AS EmployeeID, Employees.LastName AS LASTNAME, COUNT(*) AS NOrders FROM [Orders]
JOIN Employees ON Orders.EmployeeID = Employees.EmployeeID
GROUP BY Orders.EmployeeID)

Ans = Peacock

#### 3. What product was ordered the most by customers in Germany?

SELECT PID, PName, MAX(NOrders) 
FROM (SELECT Products.ProductID AS PID, Customers.CustomerID AS CID, Products.ProductName AS PName, COUNT(*) AS NOrders
FROM [Customers]
JOIN Orders ON Customers.CustomerID = Orders.CustomerID
JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
JOIN Products ON OrderDetails.ProductID = Products.ProductID
WHERE Country == 'Germany'
GROUP BY Products.ProductID)

Ans = Gorgonzola Telino